---
title: "Anova Notebook"
output:
  html_document:
    df_print: paged
---

```{r init libraries, include=FALSE}
library(data.table)
library(summarytools)
library(ggplot2)
```

## Additional information
**tRNA:** folded RNA, consists of a stem and anticodon loop. 

**aaRS:** aminoacyl-tRNA-synthetase. This enzyme is responsible for loading tRNAs with amino acids. A protein/enzyme is a folded chain of amino acids. The aaRS recognizes its cognate tRNA by its stem and anticodon.

**mRNA:** Consists of codons. 

**Ribosome:** In the ribosome the tRNA anticodon matches with a fitting mRNA codon and releases its amino acid so that an amino acid chane can be build.

**"|"** is used as an "or"


##Simulated combinations
The simulated biological world consists of the following components at the start of the simulation (the names used for the parameters are in brackets):

- **17 | 20 | 23** amino acids (initAaNumb)

- **5 | 10 | 20 | 30** mRNAs/genes (geneNumb)

- **5 | 10 | 20 | 30** aaRSs (livingAarsStartNumb)


The components have the following properties:

- aaRSs have similar functions when they have similar amino acid sequences: **true | false** (similarAars)

- aaRSs have a lifespan of **2 | 5 | 10 | 15** simulation steps (lifeticks)


Randomness is integrated by following factors:

- Each combination of initAaNumb and similarAars has a new randomly defined aaRS pool. From this pool the livingAarss are taken. Each of the combinations above was simulated with **5** different seed values for the livingAarss.

- Of these combinations each was simulated with **10** different seed values for the randomnes in the simulation itself.


**Summary:** 6 parameters with all in all 1920 combinations of factor levels and 10 observations per combination. An observation is a fitness value that is the mean of 100000 simulation steps under the conditions given by the parameters.

```{r 6 params + pre conditions, eval=FALSE, include=FALSE}
 # config_64_172023aa_ 10outSeed_5livingSeed.csv
 reducedConfig <- newConfig
 reducedConfig$aarsPath <- NULL
 reducedConfig$mrnaPath <- NULL
 reducedConfig$livingAarsPath <- NULL
 reducedConfig$translationType <- NULL
 reducedConfig$codonNumb <- NULL
 reducedConfig$mrnaSeed <- NULL
 reducedConfig$maxAnticodonNumb <- NULL
 reducedConfig$livingAarsDir <- NULL
 reducedConfig$outputSeed <- NULL
 reducedConfig <- unique(reducedConfig)
 
 summarizeByOutputSeed <- function (p1, p2, p3, p4, p5, p6){
   config$mean[which(geneNumb == p1 & initAaNumb == p2 & similarAars == p3 & aarsLifeticksStartValue == p4 & livingAarsStartNumb == p5 & livingAarsSeed == p6)]
 }
 
 outputMeansOrdered <- sapply(1:nrow(reducedConfig), FUN = function(pos){i <- reducedConfig[pos]; summarizeByOutputSeed(i[[1]], i[[2]], i[[3]], i[[4]], i[[5]], i[[6]])})
 
 outputMeansOrdered <- split(as.vector(outputMeansOrdered), seq(nrow(reducedConfig)))
 
 meansOfOutputMeans <- sapply(1: length(outputMeansOrdered), FUN = function(pos){mean(outputMeansOrdered[[pos]])})
 normalityTestPValue <- sapply(1: length(outputMeansOrdered), FUN = function (pos) {(shapiro.test(outputMeansOrdered[[pos]]))$p.value})
 
```

```{r ANOVA}
load("G:/Hanna/R/ANOVA/config.RData") 
config <- configSave

anvSet1 <- config[which(geneNumb == 30 & similarAars == 0 & aarsLifeticksStartValue == 10 & livingAarsStartNumb == 30)]

anovaResult <- aov(anvSet1$mean ~ initAaNumb, data = anvSet1)
summary(anovaResult)
plot(anovaResult$residuals, col = anvSet1$initAaNumb)
plot(anovaResult, col = anvSet1$initAaNumb)

anvSet2 <- config[which(livingAarsSeed == 20)]
colourPlots <- list(config, anvSet2)

```

```{r coloured plots}
for(pos in  1 : length(colourPlots)){
  colourPlot = colourPlots[[pos]]
  
 initAaNumb <- factor(colourPlot$initAaNumb)
 plot(colourPlot$mean, col = initAaNumb)
 legend('topright', legend = levels(initAaNumb), col = 1:6, cex = 0.8, pch = 1, title = "initAaNumb")
 
 similarAars <- factor(colourPlot$similarAars)
 plot(colourPlot$mean, col = similarAars)
 legend('topright', legend = levels(similarAars), col = 1:6, cex = 0.8, pch = 1, title= "similarAars")
 
 geneNumb <- factor(colourPlot$geneNumb)
 plot(colourPlot$mean, col = geneNumb)
 legend('topright', legend = levels(geneNumb), col = 1:6, cex = 0.8, pch = 1, title = "geneNumb")
 
 livingAarsStartNumb <- factor(colourPlot$livingAarsStartNumb)
 plot(colourPlot$mean, col = livingAarsStartNumb)
 legend('topright', legend = levels(livingAarsStartNumb), col = 1:6, cex = 0.8, pch = 1, title = "livingAarsStartNumb")
 
 livingAarsSeed <- factor(colourPlot$livingAarsSeed)
 plot(colourPlot$mean, col = livingAarsSeed)
 legend('topright', legend = levels(livingAarsSeed), col = 1:6, cex = 0.8, pch = 1, title= "livingAarsSeed")
 
 aarsLifeticksStartValue <- factor(colourPlot$aarsLifeticksStartValue)
 plot(colourPlot$mean, col = aarsLifeticksStartValue)
 legend('topright', legend = levels(aarsLifeticksStartValue), col = 1:6, cex = 0.8, pch = 1, title = "aarsLifeticksStartValue")
}
 

```

```{r tidy data, eval=FALSE, include=FALSE}
newFitness <- c()
fitnessPaths <- c()
  
  #get fitness paths
  outputPath <- paste(newConfig$livingAarsDir, paste("output_s", newConfig$outputSeed, sep = ""), sep = "")
  fitnessPaths <- paste(outputPath, "codeTableFitness.csv", sep = "\\")
  
  # read fitness values from paths
  fitnessPathsLength <- length(fitnessPaths)
  fPathsToProcess <- fitnessPaths
  # maxToTake defines how many paths are read 
  maxToTake <- length(fitnessPaths)
  if (fitnessPathsLength > maxToTake){
    fPathsToProcess <- head(fitnessPaths, maxToTake) # paths that will be read
    fitnessPaths <- tail(fitnessPaths, fitnessPathsLength - maxToTake) # rest of the paths
  }
  indices <- c(1:length(fPathsToProcess))
  newFitness <- sapply(indices, FUN = function (pos){tryCatch(fread(fPathsToProcess[pos]))})
  originalFitnessData <- c(originalFitnessData, newFitness)
  
  # tidy up workspace
  rm(fitnessPaths)
  rm(fitnessPathsLength)
  rm(newFitness)
  rm(outputPath)
  rm(maxToTake)
  rm(basePath)
  rm(fPathsToProcess)
  rm(configPath)
  
  #add number of simulation steps 
  indices <- c(1:length(originalFitnessData))
  simulationSteps <- sapply(indices, FUN = function(pos){length(originalFitnessData[[pos]])})
  newConfig <- cbind(simulationSteps, newConfig) 
  rm(simulationSteps)
  
  #delete all 0.0 fitness values in fitness data
  fitness <- originalFitnessData
  fitness <- sapply(indices,FUN =  function(pos){fitness[[pos]][which(fitness[[pos]] > 0)]})
  
  #count number of generations that the cell lived
  notNullValues <- sapply(indices,FUN =  function(pos){length(fitness[[pos]])})
  newConfig <- cbind(notNullValues, newConfig)
  rm(notNullValues)
  
  #summarize fitness data
  indices <- c(1:length(fitness))
  descrStatistics <- sapply(indices, FUN = function(pos){descr(fitness[[pos]], style = "rmarkdown", stats = c("mean", "sd", "min", "q1", "med", "q3", "max", "mad", "iqr", "cv", "skewness", "se.skewness", "kurtosis"))})
  descrStatistics <- data.table(t(descrStatistics))
  colnames(descrStatistics) <- c("mean", "sd", "min", "q1", "med", "q3", "max", "mad", "iqr", "cv", "skewness", "se.skewness", "kurtosis")
  newConfig <- cbind(descrStatistics, newConfig)
  rm(newConfig)
  rm(descrStatistics)
  
  config <- rbind(config, newConfig)

  
```

```{r write descriptive statistics to file, eval=FALSE, include=FALSE}
attach(config)
statsToPrint <- data.table(mean, sd, min, q1, med, q3, max, mad, iqr, cv, skewness, se.skewness, kurtosis, notNullValues, simulationSteps)
colnames(statsToPrint) <- c("mean", "sd", "min", "q1", "med", "q3", "max", "mad", "iqr", "cv", "skewness", "se.skewness", "kurtosis", "notNullValues", "simulationSteps")
fwrite(statsToPrint, "G:/Hanna/stats_64_172023aa_ 10outSeed_5livingSeed.csv")
rm(statsToPrint)
detach(config)
```


```{r regression models, eval=FALSE, include=FALSE}
regression <- function (param, paramValues){ 
mean <- config$mean[which(config$codonNumb == 64)]
plot(param, mean)

exp.model <- lm(log(mean) ~ param)
predictedCounts <- predict(exp.model, list(param=paramValues))
lines(paramValues, predictedCounts, col = "orange", lwd = 3)

paramSquare <- param^2
quadratic.model <- lm(mean ~ param + paramSquare)
predictedCounts <- predict(quadratic.model, list(param=paramValues, paramSquare=paramValues^2))
lines(paramValues, predictedCounts, col = "red", lwd = 3)

linear.model <- lm(mean ~ param)
predictedCounts <- predict(linear.model, list(param=paramValues))
lines(paramValues, predictedCounts, col = "darkgreen", lwd = 3)

print("linear model summary")
print(summary(linear.model))
print("quadratic model summary")
print(summary(quadratic.model))
print("exponential model summary")
print(summary(exp.model))

}

#meanValues <- seq(0, 1, 0.01)

#regression(config$codonNumb)
regression(config$geneNumb[which(config$codonNumb == 64)], seq(5, 30, 5))
regression(config$initAaNumb[which(config$codonNumb == 64)], seq(17, 23, 1))
regression(config$livingAarsSeed[which(config$codonNumb == 64)], seq(20, 25, 1))
regression(config$aarsLifeticksStartValue[which(config$codonNumb == 64)], seq(2, 15, 1))
regression(config$livingAarsStartNumb[which(config$codonNumb == 64)], seq(5, 30, 5))


```


```{r ANOVA2}
anovaResult <- aov(config$mean ~ factor(config$initAaNumb) + factor(config$geneNumb) + factor(config$aarsLifeticksStartValue) + factor(config$livingAarsStartNumb) + factor(config$livingAarsSeed) + factor(config$similarAars), data = config)

anovaData <- config[which(config$livingAarsSeed == 20)]
attach(anovaData)
anovaResult <- aov(mean ~ factor(initAaNumb) + factor(geneNumb) + factor(aarsLifeticksStartValue) + factor(livingAarsStartNumb) + factor(similarAars), data = anovaData)
detach(anovaData)
# post hoc
TukeyHSD(anovaResult)
plot(anovaResult, col= anovaData$initAaNumb)

lin <- lm(mean ~ factor(initAaNumb) + factor(geneNumb) + factor(aarsLifeticksStartValue) + factor(livingAarsStartNumb) + factor(similarAars), data = anovaData)

anova(lin)
```





