---
title: "Anova Notebook"
output:
  html_document:
    df_print: paged
---

```{r init libraries, include=FALSE}
library(data.table)
library(summarytools)
library(ggplot2)
```

```{r init base variables, eval=FALSE, include=FALSE}
config <- c()
originalFitnessData <- c()
fitness <- c()
```

```{r read config data, eval=FALSE, include=FALSE}
# read config
  basePath <- "G:/Hanna/SimulationTree/selection_false/ANOVA/translMethod_affinity/"
  configPath <- paste(basePath, "config_64_172023aa_ 1120outSeed_5livingSeed.csv", sep = "")
  
  newConfig <- c()
  newConfig <- data.table::fread(configPath)
  
  # set column names 
  colnames(newConfig) <- c("translationType", "mrnaSeed", "codonNumb", "geneLength", "geneNumb", "initAaNumb", "similarAars", "maxAnticodonNumb", "aarsLifeticksStartValue", "livingAarsStartNumb", "livingAarsSeed", "outputSeed", "addToOutput", "mrnaPath", "aarsPath", "livingAarsPath", "livingAarsDir")
  
  # delete gene lengths because it is always three
  newConfig$geneLength <- NULL
  # delete addToOutput because it is not relevant for the data analysis
  newConfig$addToOutput <- NULL
  
```

```{r 6 params + pre conditions, eval=FALSE, include=FALSE}
 # config_64_172023aa_ 10outSeed_5livingSeed.csv
 reducedConfig <- newConfig
 reducedConfig$aarsPath <- NULL
 reducedConfig$mrnaPath <- NULL
 reducedConfig$livingAarsPath <- NULL
 reducedConfig$translationType <- NULL
 reducedConfig$codonNumb <- NULL
 reducedConfig$mrnaSeed <- NULL
 reducedConfig$maxAnticodonNumb <- NULL
 reducedConfig$livingAarsDir <- NULL
 reducedConfig$outputSeed <- NULL
 reducedConfig <- unique(reducedConfig)
 
 summarizeByOutputSeed <- function (p1, p2, p3, p4, p5, p6){
   config$mean[which(config$geneNumb == p1 & config$initAaNumb == p2 & config$similarAars == p3 & config$aarsLifeticksStartValue == p4 & config$livingAarsStartNumb == p5 & config$livingAarsSeed == p6)]
 }
 
 outputMeansOrdered <- sapply(1:nrow(reducedConfig), FUN = function(pos){i <- reducedConfig[pos]; summarizeByOutputSeed(i[[1]], i[[2]], i[[3]], i[[4]], i[[5]], i[[6]])})
 
 #outputMeansOrdered <- split(as.vector(outputMeansOrdered), seq(nrow(reducedConfig)))
 
 meansOfOutputMeans <- sapply(1: length(outputMeansOrdered), FUN = function(pos){mean(outputMeansOrdered[[pos]])})
 normalityTestPValue <- sapply(1: length(outputMeansOrdered), FUN = function (pos) {(shapiro.test(outputMeansOrdered[[pos]]))$p.value})
 
 reducedConfig = cbind(reducedConfig, meansOfOutputMeans)
```

```{r configSave, eval=FALSE, cache=, include=FALSE}
configSave <- config
save(configSave, file = "G:/Hanna/R/ANOVA/config.RData") 

config6Factors <- mainData
save(configSave, file = "G:/Hanna/R/ANOVA/config6Factors.RData") 
```

```{r ANOVA}
config <- load("G:/Hanna/R/ANOVA/config.RData") 
anvSet1 <- config[which(geneNumb == 30 & similarAars == 0 & aarsLifeticksStartValue == 10 & livingAarsStartNumb == 30)]

anovaResult <- aov(anvSet1$mean ~ initAaNumb, data = anvSet1)
summary(anovaResult)
plot(anovaResult$residuals, col = anvSet1$initAaNumb)
plot(anovaResult, col = anvSet1$initAaNumb)

anvSet2 <- config[which(livingAarsSeed == 20)]
colourPlots <- list(config, anvSet2)

```

```{r coloured plots}
for(pos in  1 : length(colourPlots)){
  colourPlot = colourPlots[[pos]]
  
 initAaNumb <- factor(colourPlot$initAaNumb)
 plot(colourPlot$mean, col = initAaNumb)
 legend('topright', legend = levels(initAaNumb), col = 1:6, cex = 0.8, pch = 1)
 
 similarAars <- factor(colourPlot$similarAars)
 plot(colourPlot$mean, col = similarAars)
 legend('topright', legend = levels(similarAars), col = 1:6, cex = 0.8, pch = 1)
 
 geneNumb <- factor(colourPlot$geneNumb)
 plot(colourPlot$mean, col = geneNumb)
 legend('topright', legend = levels(geneNumb), col = 1:6, cex = 0.8, pch = 1)
 
 livingAarsStartNumb <- factor(colourPlot$livingAarsStartNumb)
 plot(colourPlot$mean, col = livingAarsStartNumb)
 legend('topright', legend = levels(livingAarsStartNumb), col = 1:6, cex = 0.8, pch = 1)
 
 livingAarsSeed <- factor(colourPlot$livingAarsSeed)
 plot(colourPlot$mean, col = livingAarsSeed)
 legend('topright', legend = levels(livingAarsSeed), col = 1:6, cex = 0.8, pch = 1)
 
 aarsLifeticksStartValue <- factor(colourPlot$aarsLifeticksStartValue)
 plot(colourPlot$mean, col = aarsLifeticksStartValue)
 legend('topright', legend = levels(aarsLifeticksStartValue), col = 1:6, cex = 0.8, pch = 1)
}
 

```

```{r tidy data, eval=FALSE, include=FALSE}
newFitness <- c()
fitnessPaths <- c()
  
  #get fitness paths
  outputPath <- paste(newConfig$livingAarsDir, paste("output_s", newConfig$outputSeed, sep = ""), sep = "")
  fitnessPaths <- paste(outputPath, "codeTableFitness.csv", sep = "\\")
  
  # read fitness values from paths
  fitnessPathsLength <- length(fitnessPaths)
  fPathsToProcess <- fitnessPaths
  # maxToTake defines how many paths are read 
  maxToTake <- length(fitnessPaths)
  if (fitnessPathsLength > maxToTake){
    fPathsToProcess <- head(fitnessPaths, maxToTake) # paths that will be read
    fitnessPaths <- tail(fitnessPaths, fitnessPathsLength - maxToTake) # rest of the paths
  }
  indices <- c(1:length(fPathsToProcess))
  newFitness <- sapply(indices, FUN = function (pos){tryCatch(fread(fPathsToProcess[pos]))})
  originalFitnessData <- c(originalFitnessData, newFitness)
  
  # tidy up workspace
  rm(fitnessPaths)
  rm(fitnessPathsLength)
  rm(newFitness)
  rm(outputPath)
  rm(maxToTake)
  rm(basePath)
  rm(fPathsToProcess)
  rm(configPath)
  
  #add number of simulation steps 
  indices <- c(1:length(originalFitnessData))
  simulationSteps <- sapply(indices, FUN = function(pos){length(originalFitnessData[[pos]])})
  newConfig <- cbind(simulationSteps, newConfig) 
  rm(simulationSteps)
  
  #delete all 0.0 fitness values in fitness data
  fitness <- originalFitnessData
  fitness <- sapply(indices,FUN =  function(pos){fitness[[pos]][which(fitness[[pos]] > 0)]})
  
  #count number of generations that the cell lived
  notNullValues <- sapply(indices,FUN =  function(pos){length(fitness[[pos]])})
  newConfig <- cbind(notNullValues, newConfig)
  rm(notNullValues)
  
  #summarize fitness data
  indices <- c(1:length(fitness))
  descrStatistics <- sapply(indices, FUN = function(pos){descr(fitness[[pos]], style = "rmarkdown", stats = c("mean", "sd", "min", "q1", "med", "q3", "max", "mad", "iqr", "cv", "skewness", "se.skewness", "kurtosis"))})
  descrStatistics <- data.table(t(descrStatistics))
  colnames(descrStatistics) <- c("mean", "sd", "min", "q1", "med", "q3", "max", "mad", "iqr", "cv", "skewness", "se.skewness", "kurtosis")
  newConfig <- cbind(descrStatistics, newConfig)
  rm(newConfig)
  rm(descrStatistics)
  
  config <- rbind(config, newConfig)

  
```

```{r write descriptive statistics to file, eval=FALSE, include=FALSE}
attach(config)
statsToPrint <- data.table(mean, sd, min, q1, med, q3, max, mad, iqr, cv, skewness, se.skewness, kurtosis, notNullValues, simulationSteps)
colnames(statsToPrint) <- c("mean", "sd", "min", "q1", "med", "q3", "max", "mad", "iqr", "cv", "skewness", "se.skewness", "kurtosis", "notNullValues", "simulationSteps")
fwrite(statsToPrint, "G:/Hanna/stats_64_172023aa_ 10outSeed_5livingSeed.csv")
rm(statsToPrint)
detach(config)
```
## Additional information
**tRNA:** folded RNA, consists of a stem and anticodon loop. 

**aaRS:** aminoacyl-tRNA-synthetase. This enzyme is responsible for loading tRNAs with amino acids. A protein/enzyme is a folded chain of amino acids. The aaRS recognizes its cognate tRNA by its stem and anticodon.

**mRNA:** Consists of codons. 

**Ribosome:** In the ribosome the tRNA anticodon matches with a fitting mRNA codon and releases its amino acid so that an amino acid chane can be build.

**"|"** is used as an "or"


##Simulated combinations
The simulated biological world consists of the following components at the start of the simulation (the names used for the parameters are in brackets):
- 17 | 20 | 23 amino acids (initAaNumb)
- 5 | 10 | 20 | 30 mRNAs/genes (geneNumb)
- 5 | 10 | 20 | 30 aaRSs (livingAarsStartNumb)

The components have the following properties:
- aaRSs have similar functions when they have similar amino acid sequences: true | false (similarAars)
- aaRSs have a lifespan of 2 | 5 | 10 | 15 simulation steps (lifeticks)

Randomness is integrated by following factors:
- Each combination of initAaNumb and similarAars has a new randomly defined aaRS pool. From this pool the livingAarss are taken. Each of the combinations above was simulated with **5** different seed values for the livingAarss.
- Of these combinations each was simulated with **10** different seed values for the randomnes in the simulation itself.

Summary: 6 parameters with all in all 1920 combinations of factor levels and 10 observations per combination. An observation is a fitness value that is the mean of 100000 simulation steps under the conditions given by the parameters.

```{r regression models, eval=FALSE, include=FALSE}
regression <- function (param, paramValues){ 
mean <- config$mean[which(config$codonNumb == 64)]
plot(param, mean)

exp.model <- lm(log(mean) ~ param)
predictedCounts <- predict(exp.model, list(param=paramValues))
lines(paramValues, predictedCounts, col = "orange", lwd = 3)

paramSquare <- param^2
quadratic.model <- lm(mean ~ param + paramSquare)
predictedCounts <- predict(quadratic.model, list(param=paramValues, paramSquare=paramValues^2))
lines(paramValues, predictedCounts, col = "red", lwd = 3)

linear.model <- lm(mean ~ param)
predictedCounts <- predict(linear.model, list(param=paramValues))
lines(paramValues, predictedCounts, col = "darkgreen", lwd = 3)

print("linear model summary")
print(summary(linear.model))
print("quadratic model summary")
print(summary(quadratic.model))
print("exponential model summary")
print(summary(exp.model))

}

#meanValues <- seq(0, 1, 0.01)

#regression(config$codonNumb)
regression(config$geneNumb[which(config$codonNumb == 64)], seq(5, 30, 5))
regression(config$initAaNumb[which(config$codonNumb == 64)], seq(17, 23, 1))
regression(config$livingAarsSeed[which(config$codonNumb == 64)], seq(20, 25, 1))
regression(config$aarsLifeticksStartValue[which(config$codonNumb == 64)], seq(2, 15, 1))
regression(config$livingAarsStartNumb[which(config$codonNumb == 64)], seq(5, 30, 5))


```


```{r ANOVA2}
anovaResult <- aov(config$mean ~ factor(config$initAaNumb) + factor(config$geneNumb) + factor(config$aarsLifeticksStartValue) + factor(config$livingAarsStartNumb) + factor(config$livingAarsSeed) + factor(config$similarAars), data = config)

anovaData <- config[which(config$livingAarsSeed == 20)]
attach(anovaData)
anovaResult <- aov(mean ~ factor(initAaNumb) + factor(geneNumb) + factor(aarsLifeticksStartValue) + factor(livingAarsStartNumb) + factor(similarAars), data = anovaData)
detach(anovaData)
# post hoc
TukeyHSD(anovaResult)
plot(anovaResult, col= anovaData$initAaNumb)

lin <- lm(mean ~ factor(initAaNumb) + factor(geneNumb) + factor(aarsLifeticksStartValue) + factor(livingAarsStartNumb) + factor(similarAars), data = anovaData)

anova(lin)
```
```{r post hoc}
tuc <- tuc$`factor(initAaNumb)`
```

```{r boxplots}
boxplot(config$mean~config$initAaNumb)
```

```{r WEKA}
conf <- config

# normal data
mainData <- data.frame(geneNumb = conf$geneNumb, initAaNumb = conf$initAaNumb, similarAars = conf$similarAars, aarsLifeticksStartValue = conf$aarsLifeticksStartValue, livingAarsStartNumb = conf$livingAarsStartNumb, livingAarsSeed = conf$livingAarsSeed, mean = conf$mean)

# reduced means that the 10 results with same output seed were summarized to one mean
fwrite(mainData, "G:/Hanna/R/ANOVA/wekaReduced_64_172023aa_ 10outSeed_5livingSeed.csv")

# copy mean values into new columne and make them nominal
nominalMean <- conf$mean
nominalMean[nominalMean <= 0.1] <- 0.1
nominalMean[nominalMean <= 0.2 & nominalMean >0.1] <- 0.2
nominalMean[nominalMean <= 0.3 & nominalMean >0.2] <- 0.3
nominalMean[nominalMean <= 0.4 & nominalMean >0.3] <- 0.4
nominalMean[nominalMean <= 0.5 & nominalMean >0.4] <- 0.5
nominalMean[nominalMean <= 0.6 & nominalMean >0.5] <- 0.6
nominalMean[nominalMean <= 0.7 & nominalMean >0.6] <- 0.7
nominalMean[nominalMean <= 0.8 & nominalMean >0.7] <- 0.8
nominalMean[nominalMean <= 0.9 & nominalMean >0.8] <- 0.9
nominalMean[nominalMean <= 1.0 & nominalMean >0.9] <- 1.0
conf <- cbind(conf, nominalMean)


mainData <- data.frame(geneNumb = conf$geneNumb, initAaNumb = conf$initAaNumb, similarAars = conf$similarAars, aarsLifeticksStartValue = conf$aarsLifeticksStartValue, livingAarsStartNumb = conf$livingAarsStartNumb, livingAarsSeed = conf$livingAarsSeed, mean = conf$nominalMean)

fwrite(mainData, "G:/Hanna/R/ANOVA/wekaNominalReduced_64_172023aa_ 10outSeed_5livingSeed.csv")


# subset: reducedConfig[which(reducedConfig$geneNumb == 30 & reducedConfig$livingAarsStartNumb == 20 & reducedConfig$similarAars == 0)]
mainData <- data.frame(initAaNumb = conf$initAaNumb, aarsLifeticksStartValue = conf$aarsLifeticksStartValue, livingAarsSeed = conf$livingAarsSeed, mean = conf$nominalMean)

fwrite(mainData, "G:/Hanna/R/ANOVA/wekaNominalReducedSub1_64_172023aa_ 10outSeed_5livingSeed.csv")

# notNullValues instead of mean
mainData <- data.frame(geneNumb = conf$geneNumb, initAaNumb = conf$initAaNumb, similarAars = conf$similarAars, aarsLifeticksStartValue = conf$aarsLifeticksStartValue, livingAarsStartNumb = conf$livingAarsStartNumb, livingAarsSeed = conf$livingAarsSeed, lifespan = conf$notNullValues)

fwrite(mainData, "G:/Hanna/R/ANOVA/wekaLifespan_64_172023aa_ 10outSeed_5livingSeed.csv")


# copy notNullValues into new columne and make them nominal
nominal <- conf$notNullValues
nominal[nominal <= 2] <- 2
nominal[nominal <= 4 & nominal >2] <- 4
nominal[nominal <= 5 & nominal >4] <- 5
nominal[nominal <= 9 & nominal >5] <- 9
nominal[nominal <= 10 & nominal >9] <- 10
nominal[nominal <= 14 & nominal >10] <- 14
nominal[nominal <= 15 & nominal >14] <- 15
nominal[nominal <= 19 & nominal >15] <- 19
nominal[nominal <= 20 & nominal >19] <- 20
nominal[nominal <= 29 & nominal >20] <- 29
nominal[nominal <= 30 & nominal >29] <- 30
nominal[nominal <= 100 & nominal >30] <- 100
nominal[nominal <= 100000 & nominal >100] <- 100000
conf <- cbind(conf, nominal)

# notNullValues instead of mean
mainData <- data.frame(geneNumb = conf$geneNumb, initAaNumb = conf$initAaNumb, similarAars = conf$similarAars, aarsLifeticksStartValue = conf$aarsLifeticksStartValue, livingAarsStartNumb = conf$livingAarsStartNumb, livingAarsSeed = conf$livingAarsSeed, lifespan = conf$nominal)

fwrite(mainData, "G:/Hanna/R/ANOVA/wekaLifespanNom_64_172023aa_ 10outSeed_5livingSeed.csv")


# mean of living set
conf <- livingSet
mainData <- data.frame(geneNumb = conf$geneNumb, initAaNumb = conf$initAaNumb, similarAars = conf$similarAars, aarsLifeticksStartValue = conf$aarsLifeticksStartValue, livingAarsStartNumb = conf$livingAarsStartNumb, livingAarsSeed = conf$livingAarsSeed, mean = conf$binnedMean)

fwrite(mainData, "G:/Hanna/R/ANOVA/wekaLivingNom_64_172023aa_ 10outSeed_5livingSeed.csv")

```

```{r neurall net test}
set.seed(500)
conf <- configSave
data <- data.frame(ivingAarsSeed = conf$livingAarsSeed, geneNumb = conf$geneNumb, mean = conf$mean)
index <- sample(1:nrow(data),round(0.75*nrow(data)))
train <- data[index,]
test <- data[-index,]
lm.fit <- lm(train$mean~., data=train)
summary(lm.fit)
pr.lm <- predict(lm.fit,test)
MSE.lm <- sum((pr.lm - test$mean)^2)/nrow(test)

maxs <- apply(data, 2, max) 
mins <- apply(data, 2, min)

scaled <- as.data.frame(scale(data, center = mins, scale = maxs - mins))

train_ <- scaled[index,]
test_ <- scaled[-index,]

library(neuralnet)
n <- names(train_)
f <- as.formula(paste("mean ~", paste(n[!n %in% "mean"], collapse = " + ")))
nn <- neuralnet(f,data=train_,hidden=c(2),linear.output=T)



# Logistic Regression
# where F is a binary factor and 
# x1-x3 are continuous predictors 
fit <- glm(F~x1+x2+x3,data=mydata,family=binomial())
summary(fit) # display results
confint(fit) # 95% CI for the coefficients
exp(coef(fit)) # exponentiated coefficients
exp(confint(fit)) # 95% CI for exponentiated coefficients
predict(fit, type="response") # predicted values
residuals(fit, type="deviance") # residuals
```
** analyze number of amino acids**
The results implicate that using more amino acids in the translation process reduces the fitness.
```{r t-tests}
tTest <- function (x, y){
x.mean <- mean(x)
y.mean <- mean(y)
delta <- (x.mean-y.mean)
result <- pwr.t.test(d = delta/sd(config$mean), sig.level = 0.05, power = 0.9, type = 'two.sample')
result
n <- result$n
x.size <- length(x)
y.size <- length(y)
x
y
result <- t.test(x,y)
result
}
"17 amino acids vs. 20 amino acids"
x <- config$mean[config$initAaNumb == 17]
y <- config$mean[config$initAaNumb == 20]
tTest(x,y)
"20 amino acids vs. 23 amino acids"
x <- config$mean[config$initAaNumb == 20]
y <- config$mean[config$initAaNumb == 23]
tTest(x,y)
"17 amino acids vs. 23 amino acids"
x <- config$mean[config$initAaNumb == 17]
y <- config$mean[config$initAaNumb == 23]
tTest(x,y)
boxplot(config$mean~factor(config$initAaNumb))
means <- tapply(config$mean,config$initAaNumb,mean)
points(means,col="red",pch=18)
#lines(means)

livingSet <- config[config$notNullValues == 100000]

data <- config$mean[config$initAaNumb == 17 & config$notNullValues == 100000]
norm <- rnorm(length(data), mean(data), sd(data))
chisq.test(data, norm)

levene.test

aovResult <- aov(config$mean~factor(config$initAaNumb))
aovResult
summary(aovResult)
summary(lm(config$mean~config$initAaNumb))
TukeyHSD(aovResult)

```

```{r}
livingSet <- config[config$notNullValues == 100000]
# hist shows that the following breaks make sense
breaks = seq(0.05,0.3, 0.001)
# map the means to its bins
binned <- .bincode(livingSet, breaks = breaks)
binmap <- sapply(binned, FUN = function(x) breaks[x])
hist(binmap, breaks = breaks)

# get properties of bins
livingSet <- cbind("binnedMean" = binmap, livingSet)

```

```{r}
geneNumb<- factor(livingSet$geneNumb)
lt <- factor(livingSet$aarsLifeticksStartValue)
lSeed <-  factor(livingSet$livingAarsSeed)
lNumb <- factor(livingSet$livingAarsStartNumb)
initAa <-  factor(livingSet$initAaNumb)
similar <- factor(livingSet$similarAars)
aovResult <- aov(livingSet$mean~ geneNumb + lt + initAa + similar + lNumb + lSeed)
aovResult
aovSum <- summary(aovResult)
aovSum
aovSum[[1]]$`Pr(>F)`
tukey <- TukeyHSD(aovResult)
tukey
boxplot(tukey)
summary(aov(livingSet$mean~ geneNumb * lt + initAa + similar + lNumb + lSeed))
summary(aov(livingSet$mean~ geneNumb * initAa + lt + similar + lNumb + lSeed))
summary(aov(livingSet$mean~ geneNumb * similar + initAa + lt + lNumb + lSeed))
summary(aov(livingSet$mean~ geneNumb * lNumb + initAa + similar + lt + lSeed))
summary(aov(livingSet$mean~ geneNumb * lSeed + initAa + similar + lNumb + lt))

summary(aov(livingSet$mean~ geneNumb + lt * initAa + similar + lNumb + lSeed))
summary(aov(livingSet$mean~ geneNumb + lt + initAa * similar + lNumb + lSeed))
summary(aov(livingSet$mean~ geneNumb + lt + initAa + similar * lNumb + lSeed))
summary(aov(livingSet$mean~ geneNumb + lt + initAa + similar + lNumb * lSeed))

```


